### PR TITLE
Calculate the Unicode table atomically

### LINKED ISSUE #302: Initial calculation of unicode table is not atomic
If you're using e.g. pytest-xdist this can result in a corrupt unicode table, which is bad. Instead the file should be created under a temporary name then moved atomically.

### PR DESCRIPTION
**Pull request by alexwlchan**
At least, that’s the theory.

Fixes #302.

### CODE REVIEW

File: src/hypothesis/internal/charmap.py [hunk 1]
```diff
@@ -53,8 +55,13 @@ def charmap():
             data = sorted(
                 (k, tuple((map(tuple, v))))
                 for k, v in _charmap.items())
-            with gzip.GzipFile(f, 'wb', mtime=1) as o:
+
+            # Write the Unicode table atomically
+            _, tmpfile = tempfile.mkstemp()
+            with gzip.GzipFile(tmpfile, 'wb', mtime=1) as o:
                 o.write(pickle.dumps(data, pickle.HIGHEST_PROTOCOL))
+            shutil.move(tmpfile, f)
```
Comment (untitaker) [inline src/hypothesis/internal/charmap.py]: `os.rename` works as well.

File: src/hypothesis/internal/charmap.py [hunk 2]
```diff
@@ -53,8 +55,13 @@ def charmap():
             data = sorted(
                 (k, tuple((map(tuple, v))))
                 for k, v in _charmap.items())
-            with gzip.GzipFile(f, 'wb', mtime=1) as o:
+
+            # Write the Unicode table atomically
+            _, tmpfile = tempfile.mkstemp()
+            with gzip.GzipFile(tmpfile, 'wb', mtime=1) as o:
                 o.write(pickle.dumps(data, pickle.HIGHEST_PROTOCOL))
```
Comment (untitaker) [inline src/hypothesis/internal/charmap.py]: Maybe serialize before creating the tempfile, so we don't have to add extra code for deleting the tempfile if serialization fails.

**Comment by DRMacIver** [comment]
Thanks!

**Comment by DRMacIver** [comment]
FYI, and this is my fault for not catching it (my windows builds weren't running and I didn't notice), not your fault for introducing it, this broke Windows support. 

```
Traceback (most recent call last): 
  File "C:\Python27\lib\site-packages\_pytest\config.py", line 318, in _importconftest
    mod = conftestpath.pyimport() 
  File "C:\Python27\lib\site-packages\py\_path\local.py", line 650, in pyimport
    __import__(modname) 
  File "C:\projects\hypothesis\tests\conftest.py", line 26, in <module>
    run() 
  File "C:\projects\hypothesis\tests\common\setup.py", line 34, in run
    charmap() 
  File "C:\Python27\lib\site-packages\hypothesis\internal\charmap.py", line 64, in charmap
    shutil.move(tmpfile, f) 
  File "C:\Python27\lib\shutil.py", line 303, in move
    os.unlink(src) 
WindowsError: [Error 32] The process cannot access the file because it is being used by another process: 
```

I'll fix this, so don't worry about it, I'm just mentioning it as a point of interest.

**Pull request merged**